share/lib/python/neuron/__init__.py

```diff
@@ -632,15 +632,15 @@ def nrn_dll(printpath=False):
         be used with care.
     """
     import ctypes
-    import os
-    import platform
     import glob
+    import os
+    import sys
 
     try:
         # extended? if there is a __file__, then use that
         if printpath:
             print("hoc.__file__ %s" % _original_hoc_file)
-        the_dll = ctypes.cdll[_original_hoc_file]
+        the_dll = ctypes.pydll[_original_hoc_file]
```

Comment:
> Is it because when the user later uses the dll to
call a function, the function wrapper in ctypes will not release the GIL (because it is a pydll)

Yes, so the C/C++ function can use the Python API without explicitly acquiring the GIL.

> more often than not, isn't it the case that the function doesn't need the GIL

I'm not sure, and didn't do an exhaustive survey. I had the impression that it was "quite common" to use the C API without acquiring the GIL. For the record, https://github.com/neuronsimulator/nrn/blob/24ebec99e178393659c3181757e93c8d79388e09/src/nrnpython/nrnpy_hoc.cpp#L2609-L2634 is the method that was causing problems.

> It there is ever an issue with thread performance in the future, I guess we can add a non-blocking version.

Indeed, it seems preferable to be "safe by default" and optimise problematic cases as and when they arise.

Reply:
Yes. The only problem is when the Python API is used from the NEURON C function. Using the Python API without the GIL won't always lead to a segfault, but it can.

Most of the time the functions we call this way don't use the Python API and are perfectly fine without the GIL. You can do a lot without the Python API; e.g. `nrn_hocobj_ptr` can generate a PyObject for a pointer without invoking the API as a PyObject is just a certain type of struct. However, I realized this morning that some of the functions involved in setting up 3D intra/extracellular simulation do use the API.

I think I'd be hard-pressed to find an example where performance is an issue, because this would only come up when calling code from Python, which generally means we're not doing simulation, so we're probably running single-threaded anyways.